### PR TITLE
Implementation of the rest of the gpg signing support

### PR DESCRIPTION
**Pull request by necauqua**
So I've been using this stack locally for a while and I'm very happy with it, but I couldn't find it in me (at least for a little while at the moment) to actually finish it, and the _only_ thing that has to be done here is just more tests 🤷

And well also I just noticed the changelog checkbox, changelog too
And also SSH impl, I was focused on GPG, but an SSH impl should be trivial at this point (cc @chriskrycho)

I'm just pushing this just so that it's not only on my machine, and maybe for someone to take it and make the final push to have it in main (maybe me after a few months) - every commit is independent and can be cherry-picked (hm, except I think in the templater one I touch some stuff in the signer, could conflict a little, eh)

Also the GPG impl commit has those weird tests that need a gpg binary in path which I'm 95% certain will fail on windows/mac, and maybe on linux too :upside_down_face: 

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by necauqua** [comment]
huh codespell actually caught something, cool

**Comment by necauqua** [comment]
Huuuuuh, gpg is not only present on the windows worker, but even works (in a sense that jj successfully calls the command, so a build off this PR will work on windows) - only thing breaking seems to be how I import the test keys - idk how to fix, and I've no windows machine

And it did work on mac too

File: lib/src/gpg_signing.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            })
+        } else {
+            Ok(output.stdout)
+        }
+    }
+}
+
+impl SigningBackend for GpgBackend {
+    fn name(&self) -> &str {
+        "GPG"
```
Comment (yuja) [inline lib/src/gpg_signing.rs:100]: nit: maybe better to stick to lowercase "gpg"? I expect config key would be lowercase (with dash or underscore.)

File: lib/src/gpg_signing.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+        // Search for one of the:
+        //  [GNUPG:] GOODSIG <long keyid> <primary uid..>
+        //  [GNUPG:] EXPKEYSIG <long keyid> <primary uid..>
+        //  [GNUPG:] NO_PUBKEY <long keyid>
+        //  [GNUPG:] BADSIG <long keyid> <primary uid..>
+        // in the output from --status-fd=1
+        // Assume signature is invalid if none of the above was found
+        output
+            .split(|&b| b == b'\n')
```
Comment (yuja) [inline lib/src/gpg_signing.rs:142]: Perhaps, this text parsing part can be split to function, and added test for that.

File: lib/src/gpg_signing.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        let output = self.run(
+            data,
+            &[
+                "--status-fd=1".as_ref(),
+                "--verify".as_ref(),
+                signature_file.path().as_os_str(), /* the only reason we have those .as_refs
+                                                    * transmuting to &OsStr everywhere.. */
+                "-".as_ref(),
+            ],
```
Comment (yuja) [inline lib/src/gpg_signing.rs:130]: Maybe `run()` could be split to two functions? If the first half returned a `Command` object, we don't have to build a slice of a certain type.

File: lib/tests/test_gpg.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+use jj_lib::gpg_signing::GpgBackend;
+use jj_lib::signing::{SigStatus, SignError, SigningBackend, Verification};
+use once_cell::sync::Lazy;
+
+static GPG_HOME: Lazy<PathBuf> = Lazy::new(|| {
+    let dir = tempfile::Builder::new()
+        .prefix("jj-test-")
+        .tempdir()
+        .unwrap()
+        .into_path();
```
Comment (yuja) [inline lib/tests/test_gpg.rs:19]: Maybe you know, but this temporary directory isn't removed at exit.

File: cli/src/commands/sign.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    key: Option<String>,
+    /// What revision to sign
+    #[arg(long, short, default_value = "@")]
+    revision: RevisionArg,
+    /// Sign the commit that is not authored by you or was already signed.
+    #[arg(long, short)]
+    force: bool,
+    /// Drop the signature, explicitly "un-signing" the commit.
+    #[arg(long, short = 'D', conflicts_with = "force")]
+    drop: bool,
```
Comment (yuja) [inline cli/src/commands/sign.rs:23]: nit: might be better to make `sign --drop` a separate command. `jj` tends to have create/delete/set/get/list verbs as subcommands.

**Review by yuja — COMMENTED** [review]
(just random comments for the first two. I haven't reviewed the other patches.)

**Comment by julienvincent** [comment]
I started work on adding an ssh backend - initial wip implementation based on the GpgBackend can be found here: https://github.com/julienvincent/jujutsu/commit/eb623e618182a21e1ec3796fa53170c3adfb5fe9

This commit was created and signed by that commit! I love that haha.

File: lib/tests/test_gpg.rs (hunk 4 above)
Comment (julienvincent) [reply lib/tests/test_gpg.rs:19]: The tempfile crate implements the Drop trait which cleans up the file. This should be fine!

**Comment by julienvincent** [comment]
A comment I have after using this - I don't think commands like `jj log` and `jj show` should show or verify the signature by default.

1) Even if someone has signing enabled they probably don't have the necessary public keys for verification on their machine machine.

2) It has a not-insignificant performance impact when viewing the log - especially on larger repos.

Git only performs signature validation when the `--show-signature` flag is passed - I think `jj` should expose the same or similar behaviour.

I'll make this change on my branch unless someone objects.

File: lib/tests/test_gpg.rs (hunk 4 above)
Comment (yuja) [reply lib/tests/test_gpg.rs:19]: It's static object, so `Drop` won't be called afaik.

**Comment by necauqua** [comment]
> might not have the necessary public keys configured for verification on their local machine

Well then it'll show `?`s, which is "unknown", all part of the plan - it specifically does show you that you need to import some keys if you enabled signing

> or similar behaviour

I thought about it, but it didn't seem _that_ bad, and it's only there if you enable signing, again

We could remove it from the default template and add docs that show how to get it back if someone wants them (also check out `builtin_log_detailed`)
I had no issues, but I've never used repos that would have thousands of signatures either - and I really like the green checkmarks it gives me :)

File: lib/src/gpg_signing.rs (hunk 1 above)
Comment (julienvincent) [reply lib/src/gpg_signing.rs:100]: Agreed. Resolved in #3007

**Comment by necauqua** [comment]
Superseded by #3007 and it's followups

**Pull request closed without merging**